### PR TITLE
Conference service: REST API + Swagger UI

### PR DESCRIPTION
**Pull request by ilya-a87**

### CODE REVIEW

File: conference/src/main/java/com/kyriba/conference/api/HallController.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+
+@Api(value = "Hall management endpoint")
+@RestController
+@RequestMapping("/api/v1/halls")
+public class HallController
+{
+  private final HallService hallService;
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/HallController.java]: Implemenation is not needed at that stage.

File: conference/src/main/java/com/kyriba/conference/api/HallController.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+
+  @Autowired
+  public HallController(HallService hallService)
+  {
+    this.hallService = hallService;
+  }
+
+
+  @ApiOperation(value = "Show all conference halls", response = HallResponse.class, responseContainer = "List")
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/HallController.java]: Swagger is aware of Spring MVC annotations, so no need to provide info about response.

File: conference/src/main/java/com/kyriba/conference/api/HallController.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  List<HallResponse> findAllHalls()
+  {
+    return hallService.findAll().stream()
+        .map(HallResponse::new)
+        .collect(Collectors.toList());
+  }
+
+
+  @ApiOperation(value = "Create conference hall", response = HallResponse.class)
+  @PostMapping(produces = APPLICATION_JSON_UTF8_VALUE, consumes = APPLICATION_JSON_UTF8_VALUE)
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/HallController.java]: produces and consumes may be moved to class level.

File: conference/src/main/java/com/kyriba/conference/api/HallController.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  HallResponse createHall(@ApiParam(value = "Hall creation object", required = true) HallRequest hall)
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/HallController.java]: Usually only ID or wrapper is returned on creation.

File: conference/src/main/java/com/kyriba/conference/api/HallController.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  {
+    return new HallResponse(hallService.createHall(hall));
+  }
+
+
+  @ApiOperation(value = "Change conference hall parameters", response = HallResponse.class)
+  @PutMapping(value = "/{id}", produces = APPLICATION_JSON_UTF8_VALUE, consumes = APPLICATION_JSON_UTF8_VALUE)
+  HallResponse updateHall(@ApiParam(value = "Hall identity", required = true) @PathVariable String id,
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/HallController.java]: Required is set to true, but no validation is configured.

File: conference/src/main/java/com/kyriba/conference/api/HallController.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                          @ApiParam(value = "Hall change object", required = true) HallRequest hall)
+  {
+    return new HallResponse(hallService.updateHall(Long.valueOf(id), hall));
+  }
+
+
+  @ApiOperation(value = "Remove conference hall", response = HallResponse.class)
+  @DeleteMapping(value = "/{id}", produces = APPLICATION_JSON_UTF8_VALUE)
+  HallResponse removeHall(@ApiParam(value = "Hall identity", required = true) @PathVariable String id)
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/HallController.java]: There is no reason to return something on removal.

File: conference/src/main/java/com/kyriba/conference/api/ScheduleController.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.List;
+
+import static java.time.LocalTime.of;
+import static java.util.Arrays.asList;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+
+@Api(value = "Conference schedule management endpoint")
+@RestController
+@RequestMapping("/api/v1/schedule")
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/ScheduleController.java]: Common API prefix may be moved to common parent class.

File: conference/src/main/java/com/kyriba/conference/api/ScheduleController.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  {
+    return PresentationResponse.builder()
+        .id(55L)
+        .build();
+  }
+
+
+  @ApiOperation(value = "View presentation information", response = PresentationResponse.class)
+  @GetMapping(value = "/presentations/{id}", produces = APPLICATION_JSON_UTF8_VALUE)
+  PresentationResponse getPresentation(
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/ScheduleController.java]: The same comments as for another controller.

File: conference/src/main/java/com/kyriba/conference/api/ScheduleController.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      @ApiParam(value = "Presentation identity", required = true) @PathVariable String id)
+  {
+    return PresentationResponse.builder()
+        .id(55L)
+        .build();
+  }
+
+
+  @ApiOperation(value = "Change presentations parameters", response = PresentationResponse.class)
+  @PatchMapping(value = "/presentations/{id}", produces = APPLICATION_JSON_UTF8_VALUE, consumes = APPLICATION_JSON_UTF8_VALUE)
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/ScheduleController.java]: Don't use patch for update, use PUT instead.

File: conference/src/main/java/com/kyriba/conference/api/dto/HallRequest.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+
+@ApiModel(description="Request model for hall operation action")
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class HallRequest
+{
+  @ApiModelProperty(value = "Name or number of the hall")
+  private String name;
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/dto/HallRequest.java]: No validation configured and tested.

File: conference/src/main/java/com/kyriba/conference/api/dto/PresentationRequest.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class PresentationRequest
+{
+  @ApiModelProperty(value = "Hall identity")
+  private Long hall;
+
+  @ApiModelProperty(value = "Presentation topic")
+  private TopicRequestResponse topic;
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/api/dto/PresentationRequest.java]: Strange class name.

File: conference/src/main/java/com/kyriba/conference/configuration/SwaggerConfig.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+@EnableSwagger2
+public class SwaggerConfig
+{
+  @Bean
+  public Docket api()
+  {
+    return new Docket(DocumentationType.SWAGGER_2)
+        .directModelSubstitute(LocalTime.class, String.class)
+        .select()
+        .apis(RequestHandlerSelectors.basePackage("com.kyriba.conference.api"))
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/configuration/SwaggerConfig.java]: Provide general API info and apply filter on URL.

File: conference/src/main/java/com/kyriba/conference/domain/Hall.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  private int places;
+
+
+  public Hall(Long id)
+  {
+    this.id = id;
+  }
+
+
+  public Hall withName(String name)
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/domain/Hall.java]: Such methods are useless, builder could be used, but not so many fields in this class.

File: conference/src/main/java/com/kyriba/conference/domain/Presentation.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.ToString;
+
+import java.time.LocalTime;
+
+
+//@Entity
+@Getter
+@ToString
+@EqualsAndHashCode
+public class Presentation
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/domain/Presentation.java]: It is impossible to populate such objects. :)

File: conference/src/main/java/com/kyriba/conference/service/HallService.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
+package com.kyriba.conference.service;
+
+
+import com.kyriba.conference.api.dto.HallRequest;
+import com.kyriba.conference.domain.Hall;
+
+import java.util.List;
+
+
+public interface HallService
```
Comment (xpinjection) [inline conference/src/main/java/com/kyriba/conference/service/HallService.java]: No services are needed at that stage.

File: conference/src/test/java/com/kyriba/conference/api/HallApiTest.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doReturn;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+import static org.springframework.restdocs.restassured3.RestAssuredRestDocumentation.document;
+import static org.springframework.restdocs.restassured3.RestAssuredRestDocumentation.documentationConfiguration;
+
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@AutoConfigureMockMvc
```
Comment (xpinjection) [inline conference/src/test/java/com/kyriba/conference/api/HallApiTest.java]: Annotation is not needed.

File: conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertNotNull;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+import static org.springframework.restdocs.restassured3.RestAssuredRestDocumentation.document;
+import static org.springframework.restdocs.restassured3.RestAssuredRestDocumentation.documentationConfiguration;
+
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@AutoConfigureMockMvc
```
Comment (xpinjection) [inline conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java]: Annotation is not needed.

File: conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+public class ScheduleApiTest
```
Comment (xpinjection) [inline conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java]: Where service documentation is located?

File: conference/src/test/java/com/kyriba/conference/api/HallApiTest.java (hunk 16 above)
Comment (ilya-a87) [reply conference/src/test/java/com/kyriba/conference/api/HallApiTest.java]: ok

File: conference/src/main/java/com/kyriba/conference/api/HallController.java (hunk 1 above)
Comment (ilya-a87) [reply conference/src/main/java/com/kyriba/conference/api/HallController.java]: yes, I just mock it in HallApiTest

File: conference/src/main/java/com/kyriba/conference/api/HallController.java (hunk 2 above)
Comment (ilya-a87) [reply conference/src/main/java/com/kyriba/conference/api/HallController.java]: removed

File: conference/src/main/java/com/kyriba/conference/api/HallController.java (hunk 3 above)
Comment (ilya-a87) [reply conference/src/main/java/com/kyriba/conference/api/HallController.java]: yes, but 'findAllHalls()' and 'removeHall(...)' do not consume.

File: conference/src/main/java/com/kyriba/conference/api/HallController.java (hunk 4 above)
Comment (ilya-a87) [reply conference/src/main/java/com/kyriba/conference/api/HallController.java]: done

File: conference/src/main/java/com/kyriba/conference/api/HallController.java (hunk 5 above)
Comment (ilya-a87) [reply conference/src/main/java/com/kyriba/conference/api/HallController.java]: done

File: conference/src/main/java/com/kyriba/conference/domain/Presentation.java (hunk 14 above)
Comment (ilya-a87) [reply conference/src/main/java/com/kyriba/conference/domain/Presentation.java]: it is not used in API - just to have a view how ScheduleService should look like

File: conference/src/main/java/com/kyriba/conference/service/HallService.java (hunk 15 above)
Comment (ilya-a87) [reply conference/src/main/java/com/kyriba/conference/service/HallService.java]: yes, I just wanted to have a look how service should be mocked in tests

File: conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java (hunk 17 above)
Comment (ilya-a87) [reply conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java]: removed

File: conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java (hunk 18 above)
Comment (ilya-a87) [reply conference/src/test/java/com/kyriba/conference/api/ScheduleApiTest.java]: added index.adoc

**Review by asamal — APPROVED** [review]

**Pull request merged**